src/models/transfers.js

```diff
@@ -0,0 +1,354 @@
+  if (!transfer) {
+    throw new NotFoundError('Unknown transfer ID')
+  }
+
+  return transfer.getDataExternal()
+}
+
+function * getTransferStateReceipt (id, signatureType, conditionState) {
+  log.debug('fetching state receipt for transfer ID ' + id)
+  const transfer = yield Transfer.findById(id)
+  const transferState = transfer ? transfer.state : 'nonexistent'
+  const message = {
+    id: uri.make('transfer', id),
+    state: transferState
+  }
+  const transferStateReceipt = {
+    type: signatureType,
+    message: message,
+    signer: config.server.base_uri
+  }
+
+  if (signatureType === 'ed25519-sha512') {
+    transferStateReceipt.public_key = config.keys.ed25519.public
+    transferStateReceipt.signature = sign(hashJSON(message))
+  } else if (signatureType === 'sha256') {
+    const realPreImage = makePreImage(message.id, transferState)
+    transferStateReceipt.digest = sha256(stringifyJSON(realPreImage))
+    transferStateReceipt.message = makePreImage(message.id, transferState)
+    if (conditionState) {
+      const conditionPreImage = makePreImage(message.id, conditionState)
+      transferStateReceipt.condition_state = conditionState
+      transferStateReceipt.condition_digest = sha256(stringifyJSON(conditionPreImage))
+    }
+  } else {
+    // TODO: move this validation to controller or remove "URI"
+    throw new InvalidUriParameterError('type is not valid')
+  }
+
+  return transferStateReceipt
+}
+
+function makePreImage (transfer_id, state) {
+  return {
+    id: transfer_id,
+    state: state,
+    token: sign(sha512(transfer_id + ':' + state))
+  }
+}
+
+function sign (base64Str) {
+  return tweetnacl.util.encodeBase64(
+    tweetnacl.sign.detached(
+      tweetnacl.util.decodeBase64(base64Str),
+      tweetnacl.util.decodeBase64(config.keys.ed25519.secret)))
+}
+
+function sha256 (str) {
+  return crypto.createHash('sha256').update(str).digest('base64')
+}
+
+function sha512 (str) {
+  return crypto.createHash('sha512').update(str).digest('base64')
+}
+
+function updateTransferObject (originalTransfer, transfer) {
+  let updatedTransferData = originalTransfer.getData()
+
+  // Ignore null properties
+  updatedTransferData = _.omit(updatedTransferData, _.isNull)
+
+  // Ignore internally managed properties
+  transfer.state = updatedTransferData.state
+  transfer.created_at = updatedTransferData.created_at
+  transfer.updated_at = updatedTransferData.updated_at
+  transfer.proposed_at = updatedTransferData.proposed_at
+  transfer.prepared_at = updatedTransferData.prepared_at
+  transfer.executed_at = updatedTransferData.executed_at
+  transfer.rejected_at = updatedTransferData.rejected_at
+
+  // Ignore undefined properties
+  const transferData = _.omit(transfer.getData(), _.isUndefined)
+
+  // Clients can add authorizations
+  // The validity of these authorizations will be checked
+  // in the validateAuthorizations function
+  _.forEach(updatedTransferData.debits, function (funds, i) {
```

Comment:
This code is just moved from controllers, low-level refactoring of the copied code is out of scope for this pull request.